pkg/http/server.go

```diff
@@ -0,0 +1,245 @@
+	Handler http.Handler
+
+	// BindAddress is the address the HTTP server should listen on.
+	BindAddress string
+
+	// SecureBindAddress is the address the HTTPS server should listen on.
+	SecureBindAddress string
+
+	// TLS is the TLS configuration for the server.
+	TLS *options.TLS
+}
+
+// NewServer creates a new Server from the options given.
+func NewServer(opts Opts) (Server, error) {
+	s := &server{
+		handler: opts.Handler,
+	}
+	if err := s.setupListener(opts); err != nil {
+		return nil, fmt.Errorf("error setting up listener: %v", err)
+	}
+	if err := s.setupTLSListener(opts); err != nil {
+		return nil, fmt.Errorf("error setting up TLS listener: %v", err)
+	}
+
+	return s, nil
+}
+
+// server is an implementation of the Server interface.
+type server struct {
+	handler http.Handler
+
+	listener    net.Listener
+	tlsListener net.Listener
+}
+
+// setupListener sets the server listener if the HTTP server is enabled.
+// The HTTP server can be disabled by setting the BindAddress to "-" or by
+// leaving it empty.
+func (s *server) setupListener(opts Opts) error {
+	if opts.BindAddress == "" || opts.BindAddress == "-" {
+		// No HTTP listener required
+		return nil
+	}
+
+	networkType := getNetworkScheme(opts.BindAddress)
+	listenAddr := getListenAddress(opts.BindAddress)
+
+	listener, err := net.Listen(networkType, listenAddr)
+	if err != nil {
+		return fmt.Errorf("listen (%s, %s) failed: %v", networkType, listenAddr, err)
+	}
+	s.listener = listener
+
+	return nil
+}
+
+// setupTLSListener sets the server TLS listener if the HTTPS server is enabled.
+// The HTTPS server can be disabled by setting the SecureBindAddress to "-" or by
+// leaving it empty.
+func (s *server) setupTLSListener(opts Opts) error {
+	if opts.SecureBindAddress == "" || opts.SecureBindAddress == "-" {
+		// No HTTPS listener required
+		return nil
+	}
+
+	config := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
+		NextProtos: []string{"http/1.1"},
+	}
+	if opts.TLS == nil {
+		return errors.New("no TLS config provided")
+	}
+	cert, err := getCertificate(opts.TLS)
+	if err != nil {
+		return fmt.Errorf("could not load certificate: %v", err)
+	}
+	config.Certificates = []tls.Certificate{cert}
+
+	listenAddr := getListenAddress(opts.SecureBindAddress)
+
+	listener, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		return fmt.Errorf("listen (%s) failed: %v", listenAddr, err)
+	}
+
+	s.tlsListener = tls.NewListener(tcpKeepAliveListener{listener.(*net.TCPListener)}, config)
+	return nil
+}
+
+// Start starts the HTTP and HTTPS server if applicable.
+// It will block until the context is cancelled.
+// If any errors occur, only the first error will be returned.
+func (s *server) Start(ctx context.Context) error {
+	g, groupCtx := errgroup.WithContext(ctx)
+
+	if s.listener != nil {
+		g.Go(func() error {
+			if err := s.startServer(groupCtx, s.listener); err != nil {
+				return fmt.Errorf("error starting insecure server: %v", err)
```

Comment:
So this reminds me -- I'm tempted to remove support for our `force-https` HTTP->HTTPS redirect middleware for security reasons (since we don't have any HSTS headers set from our end to give that process the appropriate protection to prevent future MITM attacks).

But from a security profile in our current state, where do we end up if a user sets both these servers, but doesn't set the `force-https` flag? Will the whole server be served on HTTP -- and what negative consequences does that have?

Maybe I need to understand the userbase more -- but I don't see OAuth2 Proxy being the first proxy facing the internet. In my mind that should be something like nginx, which should be dealing with the HTTP->HTTPS redirects, HSTS settings (or the upstream apps do this...?).

Reply:
> But from a security profile in our current state, where do we end up if a user sets both these servers, but doesn't set the force-https flag? Will the whole server be served on HTTP -- and what negative consequences does that have?

To do this, they would have to use the alpha config for the moment. I think it's still useful to have the redirect even though we don't necessarily have the most secure HTTPS settings in that case, it's an easy way for users to enforce HTTPS without having to worry about how that's configured in another proxy.

> Will the whole server be served on HTTP -- and what negative consequences does that have?

I don't think there's any more danger than someone just running it on HTTP anyway, I think it's up to the users to make sure it's secure